target-chains/near/receiver/src/governance.rs

```diff
@@ -34,162 +29,221 @@ use {
+    /// deserialize the payload. The use of `nom` gives us parser safety, error handling, full
+    /// buffer consumption, and a more readable implementation while staying efficient.
+    pub fn deserialize(input: impl AsRef<[u8]>) -> Result<Self, Error> {
+        use nom::{
+            bytes::complete::take,
+            combinator::all_consuming,
+            multi::many0,
+            number::complete::{
+                be_u16,
+                be_u32,
+                be_u64,
+                be_u8,
+            },
+        };
+
+        let input = input.as_ref();
+
+        // Verify Governance header is as expected so we can bail to avoid more parsing.
+        let (input, magic) = take(4usize)(input)?;
+        if magic != GOVERNANCE_MAGIC {
+            return Err(Error::InvalidGovernanceModule);
         }
-    }
 
-    pub fn deserialize(data: &[u8]) -> Result<Self, Error> {
-        let mut cursor = std::io::Cursor::new(data);
-        let magic = cursor.read_u32::<BigEndian>()?;
-        let module = cursor.read_u8()?;
-        let action = cursor.read_u8()?.try_into()?;
-        let target = cursor.read_u16::<BigEndian>()?;
-
-        assert!(module == GOVERNANCE_MODULE);
-        assert!(target == 0 || target == u16::from(WormholeChain::Near));
-        assert!(magic == u32::from_le_bytes(GOVERNANCE_MAGIC));
-
-        Ok(match action {
-            ActionId::ContractUpgrade => {
-                let mut hash = [0u8; 32];
-                cursor.read_exact(&mut hash)?;
-                Self::ContractUpgrade(hash)
-            }
+        // Verify Target is this contract.
+        let (input, target) = be_u8(input)?;
+        if target != GOVERNANCE_MODULE {
+            return Err(Error::InvalidGovernanceModule);
+        }
 
-            ActionId::SetDataSources => {
-                let mut sources = Vec::new();
-                let count = cursor.read_u8()?;
-
-                for _ in 0..count {
-                    let mut emitter = [0u8; 32];
-                    cursor.read_exact(&mut emitter)?;
-                    sources.push(Source {
-                        emitter,
-                        pyth_emitter_chain: Chain::from(WormholeChain::from(
-                            cursor.read_u16::<BigEndian>()?,
-                        )),
-                    });
-                }
+        // Safely parse the action ID. [ref:action_discriminants]
+        let (input, action) = be_u8(input)?;
+        let action = GovernanceActionId::from_u8(action).ok_or(Error::InvalidGovernanceAction)?;
+
+        match action {
+            GovernanceActionId::UpgradeContract => {
+                let (_input, bytes) = all_consuming(take(32usize))(input)?;
+                let mut codehash = [0u8; 32];
+                codehash.copy_from_slice(bytes);
 
-                Self::SetDataSources(sources)
+                // Make sure the buffer has been fully consumed.
```

Comment:
this comment applies to the all_consuming above?

Reply:
Yeah some of this parsing needed cleaning up before being reviewed, was fixing up tests. Should be correct now.